src/HexagonOptimize.cpp

```diff
@@ -1118,6 +1201,221 @@ class OptimizePatterns : public IRMutator {
     }
 };
 
+class VectorReducePatterns : public IRMutator {
+    using IRMutator::visit;
+
+    // Match stencil patterns like reduction of a shuffle with below indices:
+    // 0, 1, 2,..., window_size - 1,
+    // 1, 2, 3,..., window_size,
+    // 2, 3, 4,..., window_size + 1
+    // .....
+    // window_size != lanes
+    // TODO: Their could be other patterns as well which we should match
+    int is_stencil_reduction(const Expr &op, int window_size) {
```

Comment:
I think this should be named `is_stencil_interleave` (this isn't a reduction, it's an interleave of overlapping vectors with stride 1 between them).

Also, it should be static or a free function.